File: 2021/solutions/src/bin/day06.rs
use std::fs;

fn next_day(arr: [usize; 9]) -> [usize; 9]{
    return [
        arr[1],
        arr[2],
        arr[3],
        arr[4],
        arr[5],
        arr[6],
        arr[7] + arr[0],
        arr[8],
        arr[0],
    ];
}

fn main(){
    println!("Starting...");

    let file_name = "./input/day06.txt";

    let input_file : String = fs::read_to_string(file_name)
                        .unwrap_or_else(|err| {
                            println!("Cannot read \"{file_name}\"");
                            println!("{err}");
                            std::process::exit(1);
                        });

    let mut input_lines = input_file
        .trim()
        .split(|x| x == '\n');

    let first_line = input_lines.next().expect("Can't parse input");

    let mut input_fish = first_line.trim().split(|x| x == ',');

    let mut fish_timer = [0usize; 9];

    while let Some(fish) = input_fish.next(){
        let fish = fish.parse::<usize>().expect("cannot parse input number");
        match fish {
            0..=8 => {
                fish_timer[fish] += 1;
            }
            _ => {
                eprintln!("fish with invalid life cycle in input: {fish}");
                std::process::exit(1);
            }
        }
    }

    let starting_fish_timer = fish_timer.clone();

    for _ in 0..80 {
        fish_timer = next_day(fish_timer);
    }

    let result : usize = fish_timer.iter().sum();

    println!("Result part 1: {result}");

    fish_timer = starting_fish_timer.clone();

    for _ in 0..256 {
        fish_timer = next_day(fish_timer);
    }

    let result : usize = fish_timer.iter().sum();

    println!("Result part 2: {result}");


    //println!("Result part 1: {result}");

}

File: 2020/day06.js
const inputFileName = "input/day06.txt";
const file          = Bun.file(inputFileName);
const inputContent  = await file.text();
const inputLines    = inputContent.trim().split("\n");
const nLines        = inputLines.length;

// Part 1

let result = 0;

let currentSet = new Set();

for (const line of inputLines) {
    if (line.trim() === "") {
        result += currentSet.size;
        currentSet.clear();
    } else {
        for(const char of line) {
            currentSet.add(char);
        }
    }
}
result += currentSet.size;
currentSet.clear();

console.log(`Result part 1: ${result}`);

// Part 2

result = 0;

let currentSetArray = new Array();

for (const line of inputLines) {
    if (line.trim() === "") {
        let intersectionSet = currentSetArray.at(0);
        
        for (let i = 1; i < currentSetArray.length; i++){
            intersectionSet = intersectionSet.intersection(
                currentSetArray.at(i)
            );
        }
        
        result += intersectionSet.size;
        
        currentSetArray = new Array();
    } else {
        let currentSet = new Set();
        for(const char of line) {
            currentSet.add(char);
        }
        currentSetArray.push(currentSet);
    }
}
let intersectionSet = currentSetArray.at(0);
for (let i = 1; i < currentSetArray.length; i++){
    intersectionSet = intersectionSet.intersection(
        currentSetArray.at(i)
    );
}
result += intersectionSet.size;
currentSetArray = new Array();

console.log(`Result part 2: ${result}`);

